third_party/VulkanTutorial/OffscreenRenderingVulkanTutorial.cpp

```diff
@@ -0,0 +1,575 @@
+
+void OffscreenRenderingVulkanTutorial::Run(uint64_t frame_count) {
+  InitVulkan();
+  MainLoop(frame_count);
+  CleanUp();
+}
+
+void OffscreenRenderingVulkanTutorial::StopAsync() {
+  absl::MutexLock stop_requested_lock{&stop_requested_mutex_};
+  stop_requested_ = true;
+}
+
+void OffscreenRenderingVulkanTutorial::InitVulkan() {
+  LOG("InitVulkan");
+  // To simplify our dependencies, we don't link to Vulkan, but we use volk instead:
+  // https://github.com/zeux/volk
+  CHECK_VK_SUCCESS(volkInitialize());
+  CreateInstance();
+  volkLoadInstance(instance_);
+  // Note that, unlike what is described at
+  // https://vulkan-tutorial.com/en/Drawing_a_triangle/Presentation/Window_surface,
+  // we don't need to create a surface, as we are performing offscreen rendering.
+  PickPhysicalDevice();
+  CreateLogicalDevice();
+  // Note that, unlike what is described at
+  // https://vulkan-tutorial.com/en/Drawing_a_triangle/Presentation/Swap_chain,
+  // we don't need to create a swap chain, as we are performing offscreen rendering.
+  // Create an image in memory instead.
+  CreateOffscreenImage();
+  CreateImageView();
+  CreateRenderPass();
+  CreateGraphicsPipeline();
+  CreateFramebuffer();
+  CreateCommandPool();
+  CreateCommandBuffer();
+  CreateFence();
+}
+
+void OffscreenRenderingVulkanTutorial::MainLoop(uint64_t frame_count) {
+  LOG("MainLoop");
+  for (uint64_t frame = 0; frame < frame_count; ++frame) {
+    absl::Time next_frame_time = absl::Now() + absl::Microseconds(16667);
+    DrawFrame();
+
+    {
+      absl::MutexLock stop_requested_lock{&stop_requested_mutex_};
+      if (stop_requested_mutex_.AwaitWithDeadline(absl::Condition(&stop_requested_),
+                                                  next_frame_time)) {
+        break;
+      }
+    }
+  }
+  vkDeviceWaitIdle(device_);
+}
+
+void OffscreenRenderingVulkanTutorial::CleanUp() {
+  LOG("CleanUp");
+  vkDestroyFence(device_, fence_, nullptr);
+  // Command buffers will be automatically freed when their command pool is destroyed, so we don't
+  // need an explicit cleanup.
+  vkDestroyCommandPool(device_, command_pool_, nullptr);
+  vkDestroyFramebuffer(device_, framebuffer_, nullptr);
+  vkDestroyPipeline(device_, graphics_pipeline_, nullptr);
+  vkDestroyPipelineLayout(device_, pipeline_layout_, nullptr);
+  vkDestroyRenderPass(device_, render_pass_, nullptr);
+  vkDestroyImageView(device_, image_view_, nullptr);
+  vkFreeMemory(device_, memory_, nullptr);
+  vkDestroyImage(device_, image_, nullptr);
+  vkDestroyDevice(device_, nullptr);
+  vkDestroyInstance(instance_, nullptr);
+}
+
+void OffscreenRenderingVulkanTutorial::CreateInstance() {
+  VkApplicationInfo app_info{
+      .sType = VK_STRUCTURE_TYPE_APPLICATION_INFO,
+      .pApplicationName = "VulkanTutorial",
+      .applicationVersion = VK_MAKE_VERSION(1, 0, 0),
+      .pEngineName = "",
+      .engineVersion = VK_MAKE_VERSION(1, 0, 0),
+      .apiVersion = VK_API_VERSION_1_0,
```

Comment:
Should this be 1.1, as our headers are that version?

Reply:
I guess it makes sense, done.